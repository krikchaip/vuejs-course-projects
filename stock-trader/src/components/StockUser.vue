<template>
  <div class="card">
    <div class="card-header py-2 text-white bg-info">
      <strong>{{ name }}</strong>
      <small> [ Price: {{ price }}, Quantity: {{ quantity }} ]</small>
    </div>
    <div class="card-body form-row">
      <input
        v-model="sellQuantity"
        class="form-control form-control-sm col-5 mr-auto"
        :class="{ 'is-invalid': shouldWarn }"
        placeholder="Quantity"
        type="number"
        min="1"
      />
      <button
        class="btn btn-outline-success"
        :disabled="shouldDisable">
        Sell
      </button>
    </div>
  </div>
</template>

<script>
  export default {
    props: {
      name: String,
      price: Number,
      quantity: Number
    },
    data: () => ({
      sellQuantity: null
    }),
    computed: {
      sellQuantityNotBlank() { return !!this.sellQuantity },
      sellQuantityZeroOrNegative() { return Number(this.sellQuantity) < 1 },
      withinOwnedAmount() { return Number(this.sellQuantity) <= this.quantity },

      shouldWarn() {
        return this.sellQuantityNotBlank
          && this.sellQuantityZeroOrNegative
          || !this.withinOwnedAmount
      },

      shouldDisable() {
        return !this.sellQuantityNotBlank
          || this.sellQuantityZeroOrNegative
          || !this.withinOwnedAmount
      }
    }
  }
</script>
