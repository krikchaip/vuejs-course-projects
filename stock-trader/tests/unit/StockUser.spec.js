import { shallowMount } from '@vue/test-utils'
import StockUser from '@/components/StockUser'

describe('User Interface', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(StockUser)
  })

  it('should have quantity input', () => {
    expect(wrapper.find('input[type="number"]').exists()).toBeTruthy()
  })

  it('should have sell button', () => {
    expect(wrapper.find('button').exists()).toBeTruthy()
  })

  describe('given data containing name, price and quantity', () => {
    const data = {
      name: 'BMD',
      price: 50,
      quantity: 10
    }

    it('should display on top when passing as prop', () => {
      wrapper.setProps(data)

      const top = wrapper.find('.card-header')
      expect(top.html()).toMatchSnapshot()
    })
  })
})

describe('Sell Stock', () => {
  describe('quantity field value', () => {
    let wrapper, quantityField

    beforeEach(() => {
      wrapper = shallowMount(StockUser)
      quantityField = wrapper.find('input[type="number"]')

      wrapper.setProps({ name: 'TEST', price: 50, quantity: 10 })
    })

    describe('empty', () => {
      beforeEach(() => {
        quantityField.setValue(null)
      })

      it('sell button should be disabled', () => {
        const sellButton = wrapper.find('button')
        expect(sellButton.element.disabled).toBeTruthy()
      })

      it('quantity input box should not warn', () => {
        expect(quantityField.classes()).not.toContain('is-invalid')
      })
    })

    describe('zero', () => {
      beforeEach(() => {
        quantityField.setValue(0)
      })

      it('sell button should be disabled', () => {
        const sellButton = wrapper.find('button')
        expect(sellButton.element.disabled).toBeTruthy()
      })

      it('quantity input box should warn', () => {
        expect(quantityField.classes()).toContain('is-invalid')
      })
    })

    describe('negative', () => {
      beforeEach(() => {
        quantityField.setValue(-1)
      })

      it('sell button should be disabled', () => {
        const sellButton = wrapper.find('button')
        expect(sellButton.element.disabled).toBeTruthy()
      })

      it('quantity input box should warn', () => {
        expect(quantityField.classes()).toContain('is-invalid')
      })
    })

    describe('within the amount', () => {
      beforeEach(() => {
        quantityField.setValue(10)
      })

      it('sell button should be clickable', () => {
        const sellButton = wrapper.find('button')
        expect(sellButton.element.disabled).toBeFalsy()
      })

      it('quantity input box should not warn', () => {
        expect(quantityField.classes()).not.toContain('is-invalid')
      })
    })

    describe('more than the amount', () => {
      beforeEach(() => {
        quantityField.setValue(99)
      })

      it('sell button should be disable', () => {
        const sellButton = wrapper.find('button')
        expect(sellButton.element.disabled).toBeTruthy()
      })

      it('quantity input box should warn', () => {
        expect(quantityField.classes()).toContain('is-invalid')
      })
    })
  })
})
