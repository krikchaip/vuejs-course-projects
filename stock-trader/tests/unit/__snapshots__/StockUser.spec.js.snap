// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Interface given data containing name, price and quantity should display on top when passing as prop 1`] = `
<div class="card-header py-2 text-white bg-info"><strong>BMD</strong>
  <small> [ Price: 50, Quantity: 10 ]</small>
</div>
`;
